pkg/logger/progress_logger.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2022 VMware, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logger
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	pb "github.com/cheggaaa/pb/v3"
+	"github.com/go-logr/logr"
+	regv1 "github.com/google/go-containerregistry/pkg/v1"
+)
+
+// ProgressLogger Progress bar
+type ProgressLogger interface {
+	Start(ctx context.Context, progress <-chan regv1.Update)
+	End()
+}
+
+// NewProgressBar constructor to build a ProgressLogger responsible for printing out a progress bar using updates when
+// writing to a registry via ggcr
+func NewProgressBar(logger logr.Logger, finalMessage, errorMessagePrefix string, writer io.Writer) ProgressLogger {
+	return &ProgressBarLogger{logger: logger, finalMessage: finalMessage, errorMessagePrefix: errorMessagePrefix, writer: writer}
+}
+
+// NewNoopProgressBar constructs a Noop Progress bar that will not display anything
+func NewNoopProgressBar(logger logr.Logger) ProgressLogger {
+	return &ProgressBarNoTTYLogger{logger: logger}
+}
+
+// ProgressBarLogger display progress bar on output
+type ProgressBarLogger struct {
+	cancelFunc         context.CancelFunc
+	bar                *pb.ProgressBar
+	logger             logr.Logger
+	finalMessage       string
+	errorMessagePrefix string
+	writer             io.Writer
+}
+
+// Start the display of the Progress Bar
+func (l *ProgressBarLogger) Start(ctx context.Context, progressChan <-chan regv1.Update) {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	l.cancelFunc = cancelFunc
+	// Add a new empty line to separate the progress bar from prior output
+	fmt.Println()
+	l.bar = pb.New64(0)
+	l.bar.Set(pb.Bytes, true)
+	l.bar.Set(pb.SIBytesPrefix, true)
+	l.bar.SetWriter(l.writer)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case update := <-progressChan:
+				if update.Error != nil {
+					l.logger.Error(update.Error, l.errorMessagePrefix)
+					continue
+				}
+
+				if update.Total == 0 {
+					return
+				}
+				if !l.bar.IsStarted() {
+					l.bar.SetTotal(update.Total)
+					l.bar.Start()
+				}
+				l.bar.SetCurrent(update.Complete)
+				l.bar.Write()
+			}
+		}
+	}()
+}
+
+// End stops the progress bar and writes the final message
+func (l *ProgressBarLogger) End() {
+	if l.cancelFunc != nil {
+		l.cancelFunc()
+	}
+	l.bar.Finish()
+	l.logger.Info(l.finalMessage)
+}
+
+// ProgressBarNoTTYLogger does not display the progress bar
+type ProgressBarNoTTYLogger struct {
```

Comment:
Make the noop progress bar logger simpler. Remove `finalMessage` and also logger. Its meant to sub in as something simple for higher level tests